### PR TITLE
Basic offensive indirect

### PR DESCRIPTION
**Pull request by HussainGynai**
## Description
A basic offensive indirect play

## Associated Issue
Issue #1408

### Test Case 1
1. Setup an indirect in the offensive half of the field 
2. Run the play
3. Observe how well it works
4. Repeat with other areas on the field

### CODE REVIEW

File: soccer/gameplay/plays/restarts/basic_offensive_indirect.py [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+            lambda: self.subbehavior_with_name('pass').state == behavior.Behavior.State.completed and self.subbehavior_with_name('pass').state != tactics.
+            coordinated_pass.CoordinatedPass.State.timeout,
+            # Keep trying pass until timeout
+            'pass completes') # The pass is complete
+    
+    @classmethod
+    def score(cls):
+        gs = main.game_state()
+        return 10
+        # return 0 if behavior.Behavior.State.running or (
```
Comment (Jason27chan) [inline soccer/gameplay/plays/restarts/basic_offensive_indirect.py]: delete random comments

File: soccer/gameplay/plays/restarts/basic_offensive_indirect.py [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+
+    @classmethod
+    def is_restart(cls):
+        return True
+
+    def calc_pass_point(self): # Determines the position of the recievers and where to pass
+        ball = main.ball().pos # position of ball
+        num = random.randint(0,1) # a random int to choose which receiver to pass to
+        if (num == 1):
+            self.chip = True # If the reciever is the second one, it will be far away and require a chip
```
Comment (Jason27chan) [inline soccer/gameplay/plays/restarts/basic_offensive_indirect.py]: You don't have to comment every line lmao. It's in there, so you don't have to take it out, but for future reference, commenting every line is pretty extra and not always helpful

File: soccer/gameplay/plays/restarts/basic_offensive_indirect.py [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+        else:
+            self.chip = False # If the reciever is the first one, it will be close and not require a chip
+        if (ball.x < 0):
+            sign = 1 # If the ball is on the left side of the field the receivers need to be on the right
+        else:
+            sign = -1 # If the ball is on the right side of the field the receivers need to be on the left
+        if -constants.Field.Width/4 < ball.x and ball.x < constants.Field.Width/4: # If the ball is near the center then the close receiver needs to be on the same side as the ball
+            sign_sign = -1
```
Comment (Jason27chan) [inline soccer/gameplay/plays/restarts/basic_offensive_indirect.py]: pls change to helpful variable name

**Review by Jason27chan — CHANGES_REQUESTED** [review]

**Review by Jason27chan — APPROVED** [review]
You are a good person

**Comment by liquidmetal9015** [comment]
I'm getting:

INFO:root:Current play 'Stopped' no longer enabled, aborting
INFO:root:Chose new play: 'BasicIndirect'
WARNING:root:Ambiguous fsm transitions from state'State.preparing'.  The following states are reachable now: [<State.kicking: 2>, <State.failed: 4>];  Proceeding by taking the first option.
INFO:root:Current play 'BasicIndirect' finished running
INFO:root:Chose new play: 'BasicIndirect'
INFO:root:Current play 'BasicIndirect' no longer enabled, aborting

without the ball actually being kicked? The ambiguous transition is an issue.

Also Coordinated pass is failling:
    CoordinatedPass::failed
        LineKick::kick[robot=5]
            avoidBall(0.1)
            avoidBall(-1)
            avoidBall(-1)
            avoidBall(-1)
        rcv_pt=Point(0, 6.75)
        timeout=8.98

I don't know what that's about. It's possible I'm testing this wrong, also that we should have a better method of testing restarts, idk.

File: soccer/gameplay/plays/restarts/basic_offensive_indirect.py [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+
+import standard_play
+import skills.line_kick
+import tactics.coordinated_pass
+import enum
+import skills.move
+import random
+import situational_play_selection
+
+# A basic play for the Offensive Indect kick
```
Comment (Jason27chan) [inline soccer/gameplay/plays/restarts/basic_offensive_indirect.py]: Change # to ## for javadocs format so doxygen reads it

File: soccer/gameplay/plays/restarts/basic_offensive_indirect.py [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+        gs = main.game_state()
+        return 0 if behavior.Behavior.State.running or (
+            gs.is_ready_state()
+            and gs.is_our_indirect_kick()) else float("inf")
+
+    @classmethod
+    def is_restart(cls):
+        return True
+
+    def calc_pass_point(
```
Comment (Jason27chan) [inline soccer/gameplay/plays/restarts/basic_offensive_indirect.py:96]: Add javadocs

File: soccer/gameplay/tactics/coordinated_pass.py [hunk 6]
```diff
@@ -217,6 +217,7 @@ def ball_heading_to_receive_point(self):
         return True
 
     # Funciton that decides if an opponent robot is in the way of passing
```
Comment (Jason27chan) [inline soccer/gameplay/tactics/coordinated_pass.py]: Can you change # to ## here too? Thank you God bless

**Review by Jason27chan — CHANGES_REQUESTED** [review]
Looks good! just some small changes for doxygen

**Review by Jason27chan — APPROVED** [review]
LGTM

File: soccer/gameplay/plays/Clear/clear.py [hunk 7]
```diff
@@ -25,7 +25,6 @@
 # needs to have a unique name beyond that
```
Comment (chachmu) [inline soccer/gameplay/plays/Clear/clear.py:25]: image

File: soccer/gameplay/plays/restarts/basic_offensive_indirect.py [hunk 8]
```diff
@@ -0,0 +1,193 @@
...
+
+class BasicOffensiveIndirect(standard_play.StandardPlay):
+
+    _situationList = [
+        situational_play_selection.SituationalPlaySelector.Situation.OFFENSIVE_KICK
+    ] # yapf: disable
+
+    class State(enum.Enum):
+        move = 1  # Move recievers to proper postions
+        kick = 2  # Kick the ball to one of the recievers
```
Comment (chachmu) [inline soccer/gameplay/plays/restarts/basic_offensive_indirect.py:32]: *receiver

File: soccer/gameplay/plays/restarts/basic_offensive_indirect.py [hunk 9]
```diff
@@ -0,0 +1,193 @@
...
+
+
+class BasicOffensiveIndirect(standard_play.StandardPlay):
+
+    _situationList = [
+        situational_play_selection.SituationalPlaySelector.Situation.OFFENSIVE_KICK
+    ] # yapf: disable
+
+    class State(enum.Enum):
+        move = 1  # Move recievers to proper postions
```
Comment (chachmu) [inline soccer/gameplay/plays/restarts/basic_offensive_indirect.py:31]: *receiver

File: soccer/gameplay/plays/restarts/basic_offensive_indirect.py [hunk 10]
```diff
@@ -0,0 +1,193 @@
...
+
+    def __init__(self, indirect=None):
+        super().__init__(continuous=True)
+
+        for s in BasicOffensiveIndirect.State:
+            self.add_state(s, behavior.Behavior.State.running)
+
+        self.points = []  # The list that holds the receive points
+
+        self.chip = False  # Determines wether to chip or not
```
Comment (chachmu) [inline soccer/gameplay/plays/restarts/basic_offensive_indirect.py:43]: *whether

File: soccer/gameplay/plays/restarts/basic_offensive_indirect.py [hunk 11]
```diff
@@ -0,0 +1,193 @@
...
+
+    @classmethod
+    def is_restart(cls):
+        return True
+
+    ## Calcluates the two points which the recievers will move to
+    # One is a farther position that will be chipped to
+    # The other is closer and will be lineckicked to
+    def calc_pass_point(
+        self):  # Determines the position of the recievers and where to pass
```
Comment (chachmu) [inline soccer/gameplay/plays/restarts/basic_offensive_indirect.py:97]: *receivers

File: soccer/gameplay/plays/restarts/basic_offensive_indirect.py [hunk 12]
```diff
@@ -0,0 +1,193 @@
...
+        ball = main.ball().pos  # position of ball
+        self.num = random.randint(
+            0, 1)  # a random int to choose which receiver to pass to
+        if (self.num == 1):
+            self.chip = True  # If the reciever is the second one, it will be far away and require a chip
```
Comment (chachmu) [inline soccer/gameplay/plays/restarts/basic_offensive_indirect.py:102]: *receiver

**Review by chachmu — APPROVED** [review]
Looks pretty good, a few more typos in comments but meh :man_shrugging:

**Pull request merged**